File: src/mobi/hsz/idea/gitignore/codeInspection/IgnoreIncorrectEntryInspection.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 hsz Jakub Chrzanowski <jakub@hsz.mobi>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package mobi.hsz.idea.gitignore.codeInspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElementVisitor;
import mobi.hsz.idea.gitignore.IgnoreBundle;
import mobi.hsz.idea.gitignore.psi.IgnoreEntry;
import mobi.hsz.idea.gitignore.psi.IgnoreVisitor;
import mobi.hsz.idea.gitignore.util.Glob;
import org.jetbrains.annotations.NotNull;

import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * Inspection tool that checks if entry has correct form in specific according to the specific {@link IgnoreBundle.Syntax}.
 *
 * @author Jakub Chrzanowski <jakub@hsz.mobi>
 * @since 1.0
 */
public class IgnoreIncorrectEntryInspection extends LocalInspectionTool {
    /**
     * Checks if entry has correct form in specific according to the specific {@link IgnoreBundle.Syntax}.
     *
     * @param holder     where visitor will register problems found.
     * @param isOnTheFly true if inspection was run in non-batch mode
     * @return not-null visitor for this inspection
     */
    @NotNull
    @Override
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
        return new IgnoreVisitor() {
            @Override
            public void visitEntry(@NotNull IgnoreEntry entry) {
                String regex = entry.getText();
                if (IgnoreBundle.Syntax.GLOB.equals(entry.getSyntax())) {
                    regex = Glob.createRegex(regex, false);
                }

                try {
                    Pattern.compile(regex);
                } catch (PatternSyntaxException e) {
                    holder.registerProblem(entry,
                            IgnoreBundle.message("codeInspection.incorrectEntry.message", e.getDescription()));
                }
            }
        };
    }
}

File: src/main/kotlin/mobi/hsz/idea/gitignore/codeInspection/IgnoreIncorrectEntryInspection.kt
// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package mobi.hsz.idea.gitignore.codeInspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemsHolder
import mobi.hsz.idea.gitignore.IgnoreBundle
import mobi.hsz.idea.gitignore.psi.IgnoreEntry
import mobi.hsz.idea.gitignore.psi.IgnoreVisitor
import mobi.hsz.idea.gitignore.util.Glob
import java.util.regex.Pattern
import java.util.regex.PatternSyntaxException

/**
 * Inspection tool that checks if entry has correct form in specific according to the specific [ ].
 */
class IgnoreIncorrectEntryInspection : LocalInspectionTool() {

    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = object : IgnoreVisitor() {
        override fun visitEntry(entry: IgnoreEntry) {
            var regex = entry.text
            if (IgnoreBundle.Syntax.GLOB == entry.syntax) {
                regex = Glob.createRegex(regex, false)
            }

            try {
                Pattern.compile(regex)
            } catch (e: PatternSyntaxException) {
                holder.registerProblem(entry, IgnoreBundle.message("codeInspection.incorrectEntry.message", e.description))
            }
        }
    }
}
